File: client-sdk-python/amttp/services/compliance.py
"""Compliance Service — unified compliance evaluation via the Orchestrator."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService

# ── types ─────────────────────────────────────────────────────────────────────

ComplianceAction = Literal["ALLOW", "REQUIRE_INFO", "REQUIRE_ESCROW", "BLOCK", "REVIEW"]
EntityType = Literal["RETAIL", "INSTITUTIONAL", "VASP", "HIGH_NET_WORTH", "PEP", "UNVERIFIED"]
KYCLevel = Literal["NONE", "BASIC", "STANDARD", "ENHANCED"]
RiskTolerance = Literal["STRICT", "MODERATE", "RELAXED"]


class EntityProfile(BaseModel):
    address: str
    entity_type: str = "UNVERIFIED"
    kyc_level: str = "NONE"
    risk_tolerance: str = "MODERATE"
    jurisdiction: str = ""
    daily_limit_eth: float = 0
    monthly_limit_eth: float = 0
    single_tx_limit_eth: float = 0
    sanctions_checked: bool = False
    pep_checked: bool = False
    source_of_funds_verified: bool = False
    travel_rule_threshold_eth: float = 0
    total_transactions: int = 0
    daily_volume_eth: float = 0
    monthly_volume_eth: float = 0
    risk_score_cache: Optional[float] = None
    last_activity: Optional[str] = None
    created_at: str = ""
    updated_at: str = ""


class ComplianceCheck(BaseModel):
    service: str = ""
    check_type: str = ""
    passed: bool = False
    score: Optional[float] = None
    details: Optional[Dict[str, Any]] = None
    action_required: Optional[str] = None
    reason: Optional[str] = None


class EvaluateRequest(BaseModel):
    from_address: str
    to_address: str
    value_eth: float
    asset: Optional[str] = None
    chain_id: Optional[int] = None
    metadata: Optional[Dict[str, Any]] = None


class EvaluateResponse(BaseModel):
    decision_id: str = ""
    timestamp: str = ""
    from_address: str = ""
    to_address: str = ""
    value_eth: float = 0
    originator_profile: Optional[EntityProfile] = None
    beneficiary_profile: Optional[EntityProfile] = None
    checks: List[ComplianceCheck] = []
    action: str = "REVIEW"
    risk_score: float = 0
    reasons: List[str] = []
    requires_travel_rule: bool = False
    requires_sar: bool = False
    requires_escrow: bool = False
    escrow_duration_hours: int = 0
    processing_time_ms: float = 0


class DashboardStats(BaseModel):
    total_transactions: int = 0
    transactions_today: int = 0
    high_risk_count: int = 0
    blocked_count: int = 0
    pending_review: int = 0
    total_value_eth: float = 0
    avg_risk_score: float = 0
    compliance_rate: float = 0


class DashboardAlert(BaseModel):
    id: str = ""
    severity: str = "low"
    type: str = ""
    message: str = ""
    address: str = ""
    timestamp: str = ""
    acknowledged: bool = False


class TimelineDataPoint(BaseModel):
    timestamp: str = ""
    transactions: int = 0
    volume: float = 0
    risk_score: float = 0
    blocked: int = 0


class DecisionRecord(BaseModel):
    decision_id: str = ""
    timestamp: str = ""
    from_address: str = ""
    to_address: str = ""
    value_eth: float = 0
    action: str = ""
    risk_score: float = 0


# ── service ───────────────────────────────────────────────────────────────────


class ComplianceService(BaseService):
    """Unified compliance evaluation (Orchestrator)."""

    async def evaluate(self, request: EvaluateRequest) -> EvaluateResponse:
        """Evaluate a transaction for compliance."""
        resp = await self._http.post("/evaluate", json=request.model_dump(exclude_none=True))
        resp.raise_for_status()
        data = EvaluateResponse.model_validate(resp.json())
        self._events.emit("compliance:evaluated", data.model_dump())
        return data

    async def evaluate_with_integrity(
        self, request: EvaluateRequest, snapshot_hash: str
    ) -> Dict[str, Any]:
        """Evaluate with UI integrity verification."""
        payload = {**request.model_dump(exclude_none=True), "ui_snapshot_hash": snapshot_hash}
        resp = await self._http.post("/evaluate-with-integrity", json=payload)
        resp.raise_for_status()
        return resp.json()

    async def get_dashboard_stats(self) -> DashboardStats:
        """Get dashboard statistics."""
        resp = await self._http.get("/dashboard/stats")
        resp.raise_for_status()
        return DashboardStats.model_validate(resp.json())

    async def get_dashboard_alerts(
        self, *, limit: Optional[int] = None, severity: Optional[str] = None
    ) -> List[DashboardAlert]:
        """Get dashboard alerts."""
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if severity:
            params["severity"] = severity
        resp = await self._http.get("/dashboard/alerts", params=params)
        resp.raise_for_status()
        return [DashboardAlert.model_validate(a) for a in resp.json().get("alerts", [])]

    async def get_timeline_data(
        self, *, hours: Optional[int] = None, interval: Optional[str] = None
    ) -> List[TimelineDataPoint]:
        """Get timeline data for charts."""
        params: Dict[str, Any] = {}
        if hours is not None:
            params["hours"] = hours
        if interval:
            params["interval"] = interval
        resp = await self._http.get("/dashboard/timeline", params=params)
        resp.raise_for_status()
        return [TimelineDataPoint.model_validate(d) for d in resp.json().get("data", [])]

    async def get_sankey_flow(self, *, limit: Optional[int] = None) -> Dict[str, Any]:
        """Get Sankey flow data for value visualization."""
        params = {"limit": limit} if limit else {}
        resp = await self._http.get("/sankey-flow", params=params)
        resp.raise_for_status()
        return resp.json()

    async def get_profile(self, address: str) -> EntityProfile:
        """Get entity profile by address."""
        resp = await self._http.get(f"/profiles/{address}")
        resp.raise_for_status()
        return EntityProfile.model_validate(resp.json())

    async def update_profile(self, address: str, updates: Dict[str, Any]) -> EntityProfile:
        """Update entity profile."""
        resp = await self._http.put(f"/profiles/{address}", json=updates)
        resp.raise_for_status()
        data = EntityProfile.model_validate(resp.json())
        self._events.emit("profile:updated", data.model_dump())
        return data

    async def set_entity_type(self, address: str, entity_type: str) -> EntityProfile:
        """Set entity type for an address."""
        resp = await self._http.post(f"/profiles/{address}/set-type/{entity_type}")
        resp.raise_for_status()
        return EntityProfile.model_validate(resp.json())

    async def list_profiles(
        self, *, limit: Optional[int] = None, offset: Optional[int] = None
    ) -> List[EntityProfile]:
        """List all profiles."""
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        resp = await self._http.get("/profiles", params=params)
        resp.raise_for_status()
        return [EntityProfile.model_validate(p) for p in resp.json().get("profiles", [])]

    async def list_decisions(
        self,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        action: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> List[DecisionRecord]:
        """Get decision history."""
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        if action:
            params["action"] = action
        if from_date:
            params["from_date"] = from_date
        if to_date:
            params["to_date"] = to_date
        resp = await self._http.get("/decisions", params=params)
        resp.raise_for_status()
        return [DecisionRecord.model_validate(d) for d in resp.json().get("decisions", [])]

    async def get_entity_types(self) -> List[Dict[str, str]]:
        """Get available entity types."""
        resp = await self._http.get("/entity-types")
        resp.raise_for_status()
        return resp.json().get("entity_types", [])

    async def health(self) -> Dict[str, Any]:
        """Check orchestrator service health."""
        resp = await self._http.get("/health")
        resp.raise_for_status()
        return resp.json()

File: client-sdk/src/services/compliance.ts
/**
 * Compliance Service - Unified Compliance Evaluation
 * 
 * Provides access to the AMTTP Orchestrator's compliance evaluation
 * which combines ML risk scoring, sanctions screening, AML monitoring,
 * and geographic risk assessment.
 */

import { AxiosInstance } from 'axios';
import { BaseService } from './base';
import { EventEmitter } from '../events';

// ═══════════════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export type ComplianceAction = 'ALLOW' | 'REQUIRE_INFO' | 'REQUIRE_ESCROW' | 'BLOCK' | 'REVIEW';
export type EntityType = 'RETAIL' | 'INSTITUTIONAL' | 'VASP' | 'HIGH_NET_WORTH' | 'PEP' | 'UNVERIFIED';
export type KYCLevel = 'NONE' | 'BASIC' | 'STANDARD' | 'ENHANCED';
export type RiskTolerance = 'STRICT' | 'MODERATE' | 'RELAXED';

export interface EntityProfile {
  address: string;
  entity_type: EntityType;
  kyc_level: KYCLevel;
  risk_tolerance: RiskTolerance;
  jurisdiction: string;
  daily_limit_eth: number;
  monthly_limit_eth: number;
  single_tx_limit_eth: number;
  sanctions_checked: boolean;
  pep_checked: boolean;
  source_of_funds_verified: boolean;
  travel_rule_threshold_eth: number;
  total_transactions: number;
  daily_volume_eth: number;
  monthly_volume_eth: number;
  risk_score_cache?: number;
  last_activity?: string;
  created_at: string;
  updated_at: string;
}

export interface ComplianceCheck {
  service: string;
  check_type: string;
  passed: boolean;
  score?: number;
  details?: Record<string, unknown>;
  action_required?: string;
  reason?: string;
}

export interface EvaluateRequest {
  from_address: string;
  to_address: string;
  value_eth: number;
  asset?: string;
  chain_id?: number;
  metadata?: Record<string, unknown>;
}

export interface EvaluateResponse {
  decision_id: string;
  timestamp: string;
  from_address: string;
  to_address: string;
  value_eth: number;
  originator_profile: EntityProfile;
  beneficiary_profile: EntityProfile;
  checks: ComplianceCheck[];
  action: ComplianceAction;
  risk_score: number;
  reasons: string[];
  requires_travel_rule: boolean;
  requires_sar: boolean;
  requires_escrow: boolean;
  escrow_duration_hours: number;
  processing_time_ms: number;
}

export interface DashboardStats {
  total_transactions: number;
  transactions_today: number;
  high_risk_count: number;
  blocked_count: number;
  pending_review: number;
  total_value_eth: number;
  avg_risk_score: number;
  compliance_rate: number;
}

export interface DashboardAlert {
  id: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  type: string;
  message: string;
  address: string;
  timestamp: string;
  acknowledged: boolean;
}

export interface TimelineDataPoint {
  timestamp: string;
  transactions: number;
  volume: number;
  risk_score: number;
  blocked: number;
}

export interface DecisionRecord {
  decision_id: string;
  timestamp: string;
  from_address: string;
  to_address: string;
  value_eth: number;
  action: ComplianceAction;
  risk_score: number;
}

export interface DecisionListOptions {
  limit?: number;
  offset?: number;
  action?: ComplianceAction;
  from_date?: string;
  to_date?: string;
}

// ═══════════════════════════════════════════════════════════════════════════════
// SERVICE
// ═══════════════════════════════════════════════════════════════════════════════

export class ComplianceService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Evaluate a transaction for compliance
   * This is the main entry point for transaction evaluation
   */
  async evaluate(request: EvaluateRequest): Promise<EvaluateResponse> {
    const response = await this.http.post<EvaluateResponse>('/evaluate', request);
    this.events.emit('compliance:evaluated', response.data);
    return response.data;
  }

  /**
   * Evaluate with UI integrity verification
   * Binds the evaluation to a UI snapshot hash for audit trail
   */
  async evaluateWithIntegrity(
    request: EvaluateRequest,
    snapshotHash: string
  ): Promise<EvaluateResponse & { integrity_verified: boolean }> {
    const response = await this.http.post('/evaluate-with-integrity', {
      ...request,
      ui_snapshot_hash: snapshotHash,
    });
    return response.data;
  }

  /**
   * Get dashboard statistics
   */
  async getDashboardStats(): Promise<DashboardStats> {
    const response = await this.http.get<DashboardStats>('/dashboard/stats');
    return response.data;
  }

  /**
   * Get dashboard alerts
   */
  async getDashboardAlerts(
    options?: { limit?: number; severity?: string }
  ): Promise<DashboardAlert[]> {
    const response = await this.http.get<{ alerts: DashboardAlert[] }>('/dashboard/alerts', {
      params: options,
    });
    return response.data.alerts;
  }

  /**
   * Get timeline data for charts
   */
  async getTimelineData(
    options?: { hours?: number; interval?: string }
  ): Promise<TimelineDataPoint[]> {
    const response = await this.http.get<{ data: TimelineDataPoint[] }>('/dashboard/timeline', {
      params: options,
    });
    return response.data.data;
  }

  /**
   * Get Sankey flow data for value visualization
   */
  async getSankeyFlow(options?: { limit?: number }): Promise<{
    nodes: Array<{ id: string; name: string; category: string }>;
    links: Array<{ source: string; target: string; value: number }>;
  }> {
    const response = await this.http.get('/sankey-flow', { params: options });
    return response.data;
  }

  /**
   * Get entity profile by address
   */
  async getProfile(address: string): Promise<EntityProfile> {
    const response = await this.http.get<EntityProfile>(`/profiles/${address}`);
    return response.data;
  }

  /**
   * Update entity profile
   */
  async updateProfile(
    address: string,
    updates: Partial<EntityProfile>
  ): Promise<EntityProfile> {
    const response = await this.http.put<EntityProfile>(`/profiles/${address}`, updates);
    this.events.emit('profile:updated', response.data);
    return response.data;
  }

  /**
   * Set entity type for an address
   */
  async setEntityType(address: string, entityType: EntityType): Promise<EntityProfile> {
    const response = await this.http.post<EntityProfile>(
      `/profiles/${address}/set-type/${entityType}`
    );
    return response.data;
  }

  /**
   * List all profiles
   */
  async listProfiles(options?: { limit?: number; offset?: number }): Promise<EntityProfile[]> {
    const response = await this.http.get<{ profiles: EntityProfile[] }>('/profiles', {
      params: options,
    });
    return response.data.profiles;
  }

  /**
   * Get decision history
   */
  async listDecisions(options?: DecisionListOptions): Promise<DecisionRecord[]> {
    const response = await this.http.get<{ decisions: DecisionRecord[] }>('/decisions', {
      params: options,
    });
    return response.data.decisions;
  }

  /**
   * Get available entity types
   */
  async getEntityTypes(): Promise<Array<{ type: EntityType; description: string }>> {
    const response = await this.http.get('/entity-types');
    return response.data.entity_types;
  }

  /**
   * Check service health
   */
  async health(): Promise<{
    status: string;
    service: string;
    connected_services: Record<string, string>;
    profiles_loaded: number;
  }> {
    const response = await this.http.get('/health');
    return response.data;
  }
}
